pkg/api/validate/dynamic/dynamic.go

```diff
@@ -43,7 +42,7 @@ type Subnet struct {
 
 type ServicePrincipalValidator interface {
 	ValidateServicePrincipal(ctx context.Context, clientID, clientSecret, tenantID string) error
-	SetGetToken(gt func(ctx context.Context, log *logrus.Entry, clientID, clientSecret, tenantID string, aadEndpoint, resource string) (*adal.ServicePrincipalToken, error))
+	SetGetToken(gt aad.TokenClient)
```

Comment:
Why do we need this? Can't we just create the objects with that field?

Reply:
The dynamic struct and fields are private, so when `serviceprincipal_test.go` runs on line 125 this is used to replace the `GetToken` with the mock.
The fields aren't modifiable from the `psDynamic` object.

Is there a way to directly assign the mock functions without this?

Reply:
Add an argument to `NewValidator` and `NewServicePrincipalValidator`. Then you'll be able to create them directly with your mock tokenClient in it

Reply:
I've updated them to accept the token client as an argument.